hazelcast/src/main/java/com/hazelcast/client/impl/connection/nio/ClusterConnectorServiceImpl.java

```diff
@@ -367,39 +369,47 @@ public void run() {
         private final int initialBackoffMillis;
         private final int maxBackoffMillis;
         private final double multiplier;
-        private final boolean failOnMaxBackoff;
         private final double jitter;
         private final Random random = new Random();
         private int attempt;
         private int currentBackoffMillis;
+        private int clusterConnectTimeoutSeconds;
+        private long clusterConnectDeadline;
 
         WaitStrategy(int initialBackoffMillis, int maxBackoffMillis,
-                     double multiplier, boolean failOnMaxBackoff, double jitter) {
+                     double multiplier, int clusterConnectTimeoutSeconds, double jitter) {
             this.initialBackoffMillis = initialBackoffMillis;
             this.maxBackoffMillis = maxBackoffMillis;
             this.multiplier = multiplier;
-            this.failOnMaxBackoff = failOnMaxBackoff;
+            this.clusterConnectTimeoutSeconds = clusterConnectTimeoutSeconds;
             this.jitter = jitter;
         }
 
         public void reset() {
             attempt = 0;
+            clusterConnectDeadline = Clock.currentTimeMillis() + TimeUnit.SECONDS.toMillis(clusterConnectTimeoutSeconds);
             currentBackoffMillis = Math.min(maxBackoffMillis, initialBackoffMillis);
         }
 
         public boolean sleep() {
             attempt++;
-            if (failOnMaxBackoff && currentBackoffMillis >= maxBackoffMillis) {
-                logger.warning(String.format("Unable to get live cluster connection, attempt %d.", attempt));
+            long currentTimeMillis = Clock.currentTimeMillis();
+            if (currentTimeMillis > clusterConnectDeadline) {
+                logger.warning(String.format("Unable to get live cluster connection, cluster connect timeout %d seconds is "
+                        + "reached. Attempt %d.", clusterConnectTimeoutSeconds, attempt));
                 return false;
             }
+
             //random_between
             // Random(-jitter * current_backoff, jitter * current_backoff)
             long actualSleepTime = (long) (currentBackoffMillis - (currentBackoffMillis * jitter)
                     + (currentBackoffMillis * jitter * random.nextDouble()));
 
-            logger.warning(String.format("Unable to get live cluster connection, retry in %d ms, attempt %d"
-                    + ", retry timeout millis %d cap", actualSleepTime, attempt, maxBackoffMillis));
+            actualSleepTime = Math.min(actualSleepTime, clusterConnectDeadline - currentTimeMillis);
+
+            logger.warning(String.format("Unable to get live cluster connection, retry in %d ms, attempt %d "
+                    + ", cluster connect timeout seconds %d "
+                    + ", retry timeout millis %d cap", actualSleepTime, attempt, clusterConnectTimeoutSeconds, maxBackoffMillis));
```

Comment:
can you use a more understandable word other than `cap`, e.g. `maximum retry timeout millis `